src/xml/node.cr

```diff
@@ -1,18 +1,89 @@
+require "weak_ref"
+
 class XML::Node
   LOOKS_LIKE_XPATH = /^(\.\/|\/|\.\.|\.$)/
 
-  # Creates a new node.
-  def initialize(node : LibXML::Attr*)
-    initialize(node.as(LibXML::Node*))
+  # Every Node must keep a reference to its document Node. To keep things
+  # simple, a document Node merely references itself. An unlinked node must
+  # still reference its original document Node until adopted into another
+  # document's tree (the libxml nodes keep a pointer to their libxml doc).
+  @document : Node
+
+  # The constructors allocate a XML::Node for a libxml node once, so we don't
+  # finalize a document twice for example.
+  #
+  # We store the reference into the libxml struct (_private) for documents
+  # because every a document's XML::Node lives as long as their libxml doc.
+  #
+  # However we can lose references to subtree XML::Node, so using _private would
+  # leave dangling pointers. We thus keep a cache of weak references to all
+  # nodes in the document, so we can still collect lost references, and at worst
+  # reinstantiate a XML::Node if needed.
+  protected getter! cache : Hash(LibXML::Node*, WeakRef(Node))?
```

Comment:
I left out cleaning up the hash when node references are collected:

1. we can't cleanup in a finalizer because a thread might have been interrupted while mutating the hash;
2. we don't want to iterate the hash on every `XML::Node.new` call;
3. maybe the finalizer could mark the document's cache as dirty, so the next `XML::node.new` call will cleanup (once)?